bin/propolis-server/src/lib/spec/api_spec_v0.rs

```diff
@@ -147,119 +171,144 @@ impl From<Spec> for InstanceSpecV0 {
-                    backend: backend_name.to_owned(),
-                    device: device_name.clone(),
-                })?;
-
-            builder.add_storage_device(
-                device_name,
-                Disk {
-                    device_spec: device_spec.into(),
-                    backend_spec: backend_spec.into(),
-                },
-            )?;
-        }
-
-        // Once all the devices have been checked, there should be no unpaired
-        // backends remaining.
-        if let Some(backend) = value.backends.storage_backends.keys().next() {
-            return Err(ApiSpecError::BackendNotUsed(backend.to_owned()));
-        }
-
-        // Repeat this process for network devices.
-        for (device_name, device_spec) in value.devices.network_devices {
-            let NetworkDeviceV0::VirtioNic(device_spec) = device_spec;
-            let backend_name = &device_spec.backend_name;
-            let (_, backend_spec) = value
-                .backends
-                .network_backends
-                .remove_entry(backend_name)
-                .ok_or_else(|| ApiSpecError::BackendNotFound {
-                    backend: backend_name.to_owned(),
-                    device: device_name.clone(),
-                })?;
-
-            let NetworkBackendV0::Virtio(backend_spec) = backend_spec else {
-                return Err(ApiSpecError::GuestNicInvalidBackend(device_name));
-            };
-
-            builder.add_network_device(
-                device_name,
-                Nic { device_spec, backend_spec },
-            )?;
-        }
-
-        // SoftNPU ports can have network backends, so consume the SoftNPU
-        // device fields before checking to see if the network backend list is
-        // empty.
-        #[cfg(feature = "falcon")]
-        {
-            if let Some(softnpu_pci) = value.devices.softnpu_pci_port {
-                builder.set_softnpu_pci_port(softnpu_pci)?;
-            }
-
-            if let Some(softnpu_p9) = value.devices.softnpu_p9 {
-                builder.set_softnpu_p9(softnpu_p9)?;
-            }
-
-            if let Some(p9fs) = value.devices.p9fs {
-                builder.set_p9fs(p9fs)?;
+    fn try_from(value: InstanceSpecV0) -> Result<Self, Self::Error> {
+        let mut builder = SpecBuilder::with_board(value.board);
+        let mut devices: Vec<(String, ComponentV0)> = vec![];
+        let mut storage_backends: HashMap<String, StorageBackend> =
+            HashMap::new();
+        let mut viona_backends: HashMap<String, VirtioNetworkBackend> =
+            HashMap::new();
+        let mut dlpi_backends: HashMap<String, DlpiNetworkBackend> =
+            HashMap::new();
+
+        for (name, component) in value.components.into_iter() {
+            match component {
+                ComponentV0::CrucibleStorageBackend(_)
+                | ComponentV0::FileStorageBackend(_)
+                | ComponentV0::BlobStorageBackend(_) => {
```

Comment:
i kinda wondered if it was worth having functions like `ComponentV0::is_storage_backend() -> bool` but, upon further consideration, we would only ever call them in one place, so probably not...